data/account/participationRegistry.go

```diff
@@ -0,0 +1,947 @@
+var ErrInvalidRegisterRange = errors.New("key would not be active within range")
+
+// ErrUnknownParticipationAction is used when record is given something other than the known actions.
+var ErrUnknownParticipationAction = errors.New("unknown participation action")
+
+// ErrAlreadyInserted is used when inserting a key which already exists in the registry.
+var ErrAlreadyInserted = errors.New("these participation keys are already inserted")
+
+// ErrActiveKeyNotFound is used when attempting to update an account with no active key
+var ErrActiveKeyNotFound = errors.New("no active participation key found for account")
+
+// ErrMultipleValidKeys is used when recording a result but multiple valid keys were found. This should not be possible.
+var ErrMultipleValidKeys = errors.New("multiple valid keys found while recording key usage")
+
+// ErrMultipleKeysForID this should never happen. Multiple keys with the same participationID
+var ErrMultipleKeysForID = errors.New("multiple valid keys found for the same participationID")
+
+// ErrNoKeyForID there may be cases where a key is deleted and used at the same time, so this error should be handled.
+var ErrNoKeyForID = errors.New("no valid key found for the participationID")
+
+// ParticipationRegistry contain all functions for interacting with the Participation Registry.
+type ParticipationRegistry interface {
+	// Insert adds a record to storage and computes the ParticipationID
+	Insert(record Participation) (ParticipationID, error)
+
+	// Delete removes a record from storage.
+	Delete(id ParticipationID) error
+
+	// DeleteExpired removes all records from storage which are expired on the given round.
+	DeleteExpired(round basics.Round) error
+
+	// Get a participation record.
+	Get(id ParticipationID) ParticipationRecord
+
+	// GetAll of the participation records.
+	GetAll() []ParticipationRecord
+
+	// Register updates the EffectiveFirst and EffectiveLast fields. If there are multiple records for the account
+	// then it is possible for multiple records to be updated.
+	Register(id ParticipationID, on basics.Round) error
+
+	// Record sets the Last* field for the active ParticipationID for the given account.
+	Record(account basics.Address, round basics.Round, participationType ParticipationAction) error
+
+	// Flush ensures that all changes have been written to the underlying data store.
+	Flush() error
+
+	// Close any resources used to implement the interface.
+	Close()
+}
+
+// MakeParticipationRegistry creates a db.Accessor backed ParticipationRegistry.
+func MakeParticipationRegistry(accessor db.Pair, log logging.Logger) (ParticipationRegistry, error) {
+	return makeParticipationRegistry(accessor, log)
+}
+
+// makeParticipationRegistry creates a db.Accessor backed ParticipationRegistry.
+func makeParticipationRegistry(accessor db.Pair, log logging.Logger) (*participationDB, error) {
+	if log == nil {
+		return nil, fmt.Errorf("invalid logger provided")
```

Comment:
convert this `fmt.Errorf` to `errors.New`, since you're not formatting anything in here.